### ISSUE TITLE
awaiters should be optional

### ISSUE THREAD

**Issue by mingfang** [pulumi/pulumi-kubernetes#179]
While awaiters are nice in some cases, the concept is not natural to kubernetes; Kubernetes encourages asynchronous operations leading to eventual consistency.  

Please make awaiters optional by providing a way to toggle them on/off on a case by case basis.

**Comment by mingfang** [comment]
It would be great to have both options, 1-a way to toggle awaiters at the provider level, and then 2-a way to override on a per resource basis.

I'm don't think it's a downside because users coming from kubectl are used to this behavior.  On the contrary the downside of the awaiters are that they are opinionated and are not be correct for all use cases.

**Comment by mingfang** [comment]
I would like use this tool at different layers of abstraction.  At the lowest layer it would be behavior like pure kubernetes, like kubectl apply.  Then I can opt-in to more opinionated layers such as awaiters and maybe pulumi-cloud.

**Comment by hausdorff** [comment]
> It currently waits for the endpoints to be created. This is not useful or correct since the endpoints are dynamic.

I am curious to hear more about why you believe this. The `Service` awaiters are basically an implementation of [the official documentation on debugging `Services`](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-service/). The idea is to simply present the results of the manual steps you'd have take to understand why your `Pod`s aren't getting traffic from some `Service`. It seems to me that if this were not desirable, there wouldn't be docs about it, no?

**Comment by hausdorff** [comment]
That said, @lukehoban and I had a conversation and after consideration, we do agree that this is a useful thing to do. We basically agree on the plan to have a global setting (perhaps `kubernetes:config:skipWaitForReady`) as well as an extension of `ResourceOptions` on a per-resource basis. This is pretty trivially doable for M18.

**Comment by mingfang** [comment]
When endpoints are not created, the problem will be in the pods, replica sets, or deployments. The fault will never be in the service nor can be fixed by the service. The service is a standalone resource in kubernetes. Forcing it to depend on endpoints is unnatural to kubernetes and in my option it's just plain wrong.

**Comment by hausdorff** [comment]
Yeah, I hear you, I totally get that point of view. Hopefully these upcoming opt-out flags will help you when we release M18. :)

But, FWIW, a thing we _have_ seen users do is to mis-target `Pod`s with a faulty selector, for example, especially new users. With the exception of services of type `ExternalName`, I do think that no one ever intends to target 0 pods, and that is the reason why we registered it as an error.

Thanks for the feedback, I do appreciate you pushing back on our assumptions. :)

**Comment by mingfang** [comment]
Yes I agree that the raw kubernetes resources can be mis-used and I understand you're keeping the users in mind.  In fact I built my own higher level abstraction that auto creates services and deployments while making sure the selectors are correct. 

You can see an example here https://github.com/mingfang/jenkins-seed/blob/master/httpbin.yml.
This is the python code that reads that yaml and generates kubernetes resources using kubectl https://github.com/mingfang/kube-deploy.
I've been maintaining this tool for years now but I hope to be able to move to pulumi instead.

**Comment by hausdorff** [comment]
So, more concretely related to the proposal: the main thing to decide here is what happens to `pulumi.Output` properties.

Talking with @lukehoban I tentatively suggest that we simply make `pulumi.Output`s resolve to `undefined`. I am reticent to make them `throw Error`, as that seems like something the users would want to decide.

If anyone has thoughts on this I'd be glad to hear them.

**Comment by mingfang** [comment]
After reviewing the code for the awaiters, core_service.go in particular, I think there is a lot of value in them.  Now I understand that taking them away would cause many issues in the way the rest of this provider works.  My only issue is that the core_service awaiter should not wait for endpoints.  If we can remove that then this ticket can be closed as won't fix.

**Comment by hausdorff** [comment]
@mingfang hmm, what other issues do you see this as causing? It seems to me that if we cause `pulumi.Output` values to simply be `undefined`, this would be acceptable for many use cases?

**Comment by protometa** [comment]
I was pretty surprised by the all the waiting for Kubernetes resources as well, it's definitely not the Kubernetes way, but I started to warm up to it and eventually came to depend on it. I no longer manually link my resources with common string tokens, I define them in one resource (or generate them on the fly) and have them passed up the dependency chain automatically. In that paradigm, if a service is not linking to pods there's definitely something wrong. An override would probably be useful though. I just recently had a case where I needed to refresh to do an update but the preview of the refresh was blocked by a broken service that needed the update. I ended up just skipping the preview of the refresh, but I was thinking it would be nice to skip the checks on the service somehow.

**Comment by hausdorff** [comment]
So to summarize, we've whittled @mingfang's objection down to one thing about the await logic, which is that they don't like waiting for `Endpoint`s when awaiting a `Service`.

To test the question of whether this is worthwhile, I have spent several weeks of deploying very challenging Kubernetes applications using Pulumi and paying close attention to whether this is useful The applications run the gamut, including Helm charts, YAML, hand-written Pulumi code, _etc_.

## Correctness

Across all of these applications, it has so far been universally true that when we displayed an error as a result of failing to find an `Endpoints` object, it has pointed to a legitimate bug, either in my use of the {Chart, YAML, whatever}, or a bug in the actual {Chart, YAML, whatever}. It has never _not_ been the case that this generated spurious signal.

In each of these cases I may have started by being confused, but once I figured out the problem, I was grateful to have the error.

One recent example, the CoreOS Prometheus operator by default starts many `Services` to monitor the controller manager, etcd, and the scheduler. If you boot into a Kubernetes provider that does not expose these as `Pod`s through the Kubernetes API server, these `Service`s spuriously point to nothing. I was grateful that these `Services` failed, as it alerted me to turn them off.

## Usability

The only compelling reason I can think of at this point to turn off the `Endpoints` detection is that it creates more up-front work for users trying to use (_e.g._) Helm Charts. This could inadvertently drive away users who (incorrectly!) conclude this is a Pulumi problem, rather than an app problem. This, of course, is not the outcome we wish for.

## Conclusions

So on balance, I'm still on the fence. I'd be grateful to hear other opinions. @lukehoban @joeduffy ?

**Comment by mingfang** [comment]
@hausdorff Thanks for looking into this.  I don't want to waste your time.  I've decided to move on and use a different solution.  Please feel free to close this issue.

**Comment by hausdorff** [comment]
I'm closing #184 in favor of this issue. Quoting from that OP:

> As the await logic for Pulumi Kubernetes objects becomes more complex, it has become increasingly obvious that several API objects are going to be hard to get right in all cases without additional information provided from the user.
>
> For example, it is impossible to statically infer from a Pod definition what conditions we should wait for -- is the Pod meant to complete? Run forever? If a container fails, does that mean the Pod is in a state of error, and what impact should restartPolicy have on this decision?
>
> A different idea is to allow users to specify this await logic with a variety of higher-level classes that use annotations to specify the intended semantics of the Kubernetes resources. For example, we might have a StrictPod, which has some annotation (say "pod.pulumi.io/strict": "true"), which specifies to the awaiters that it is a Pod that should be considered failed if any container crashes inside it. Similarly, we might have CompletingPod, which the awaiters would expect would complete successfully.

**Comment by lblackstone** [comment]
We're planning to add an optional toggle to disable await logic. This is intended to be an "escape hatch" in case of bugs and unhandled edge cases.

**Comment by lblackstone** [comment]
@hausdorff I took a look at the provider code, and I think there are two decent options for implementing the await logic toggle for a particular resource. (I don't think disabling awaiters completely is a good idea).

1. Update the SDK with an optional flag on all resource classes to disable await logic for that resource
2. Check for a specified annotation on the k8s resource, and disable await logic if it is present.

Thoughts?

**Comment by lblackstone** [comment]
We discussed this offline, and decided to go with option 2. This approach has a few important benefits:
1. Does not require changes to the RPC interface
2. Works nicely for resources defined natively with Pulumi, and also with YAML/JSON manifests
3. The same approach would work for configurable timeouts
4. Prior art with the autoname annotation; easy to query from the k8s apiserver
5. If needed later, we could extend the SDK with the same options, which could use the same annotations to implement the configured options
